contracts/CliptoExchange.sol

```diff
@@ -0,0 +1,242 @@
+    }
+
+    function setCliptoMinter(address _cliptoToken, address _minter) external onlyOwner {
+        ICliptoToken(_cliptoToken).setMinter(_minter);
+    }
+
+    function registerCreator(string calldata _creatorName, string calldata _jsondata) external whenNotPaused {
+        require(!_existsCreator(msg.sender), "error: creator already registered");
+        _registerCreator(msg.sender, _creatorName, _jsondata);
+    }
+
+    function updateCreator(string calldata _jsondata) external whenNotPaused {
+        require(_existsCreator(msg.sender), "error: creator is not yet registered");
+        emit CreatorUpdated(msg.sender, _jsondata);
+    }
+
+    function newRequest(
+        address _creator,
+        address _nftReceiver,
+        address _erc20,
+        uint256 _amount,
+        string calldata _jsonData
+    ) external whenNotPaused {
+        _validateRequest(_creator, _amount);
+        _checkAllowance(_erc20, msg.sender, _amount);
+        _pay(msg.sender, address(this), _erc20, _amount);
+        _newRequest(_creator, _nftReceiver, _erc20, _amount, _jsonData);
+    }
+
+    function nativeNewRequest(
+        address _creator,
+        address _nftReceiver,
+        string calldata _jsonData
+    ) external payable whenNotPaused {
+        _validateRequest(_creator, msg.value);
+        _newRequest(_creator, _nftReceiver, address(0), msg.value, _jsonData);
+    }
+
+    function deliverRequest(uint256 _requestId, string calldata _tokenURI) external nonReentrant whenNotPaused {
+        Request storage request = requests[msg.sender][_requestId];
+        require(!request.fulfilled, "error: request already fulfilled/refunded");
+
+        uint256 feeAmount = (request.amount * _feeNumer) / _feeDenom;
+        _transferPayment(feeRecipient, request.erc20, feeAmount);
+
+        uint256 paymentAmount = request.amount - feeAmount;
+        _transferPayment(msg.sender, request.erc20, paymentAmount);
+
+        address nft = creators[msg.sender];
+        uint256 nftTokenId = ICliptoToken(nft).totalSupply();
+        ICliptoToken(nft).safeMint(request.nftReceiver, _tokenURI);
+
+        request.fulfilled = true;
+        emit DeliveredRequest(msg.sender, _requestId, nftTokenId + 1);
+    }
+
+    function refundRequest(address _creator, uint256 _requestId) external nonReentrant whenNotPaused {
+        Request storage request = requests[_creator][_requestId];
+        require(request.requester == msg.sender, "error: only requester can make a refund");
+        require(!request.fulfilled, "error: request already fulfilled/refunded");
+
+        _transferPayment(msg.sender, request.erc20, request.amount);
+        request.fulfilled = true;
+
+        emit RefundedRequest(_creator, _requestId);
+    }
+
+    function rejectRequest(uint256 _requestId) external whenNotPaused {
+        Request storage request = requests[msg.sender][_requestId];
+        require(!request.fulfilled, "error: request already fulfilled/refunded");
```

Comment:
A creator rejects a request, at this point the user has already put funds into this smart contract in order to pay for it. If we set the request to fulfilled then how does the user get a refund? Seems like a fairly critical bug to me...